libmultilabel/networks/bigru.py

```diff
@@ -0,0 +1,93 @@
+    """
+    def __init__(
+        self,
+        embed_vecs,
+        num_classes,
+        rnn_dim=512,
+        rnn_layers=1,
+        dropout=0.2,
+        activation='tanh',
+        **kwargs
+    ):
+        super(BiGRU, self).__init__(embed_vecs, dropout, activation, **kwargs)
+        assert rnn_dim%2 == 0, """`rnn_dim` should be even."""
+
+        # BiGRU
+        emb_dim = embed_vecs.shape[1]
+        self.rnn = nn.GRU(emb_dim, rnn_dim//2, rnn_layers,
+                          bidirectional=True, batch_first=True)
+
+        # context vectors for computing attention
+        self.U = nn.Linear(rnn_dim, num_classes)
+        xavier_uniform_(self.U.weight)
+
+        # linear output
+        self.final = nn.Linear(rnn_dim, num_classes)
+        xavier_uniform_(self.final.weight)
+
+    def forward(self, input):
+        text, length, indices = self.sort_data_by_length(input['text'], input['length'])
+        x = self.embedding(text) # (batch_size, length, rnn_dim)
+        x = self.embed_drop(x) # (batch_size, length, rnn_dim)
+
+        packed_inputs = pack_padded_sequence(x, length, batch_first=True)
+        x, _ = self.rnn(packed_inputs)
+        x = pad_packed_sequence(x)[0]
+        x = x.permute(1, 0, 2)
+
+        x = torch.tanh(x)
+
+        # Apply per-label attention
+        alpha = torch.softmax(self.U.weight.matmul(x.transpose(1, 2)), dim=2)
+
+        # Document representations are weighted sums using the attention
+        m = alpha.matmul(x) # (batch_size, num_classes, rnn_dim)
+
+        # Compute a probability for each label
+        x = self.final.weight.mul(m).sum(dim=2).add(self.final.bias) # (batch_size, num_classes)
+
+        return {'logits': x[indices], 'attention': alpha[indices]}
+
+    def sort_data_by_length(self, data, length):
+        """Sort data by lengths. If data is not sorted before calling `pack_padded_sequence`,
+        under `enforce_sorted=False`, the setting of `use_deterministic_algorithms` must be False.
+        To keep `use_deterministic_algorithms` True, we sort the input here and return indices for
+        restoring the original order.
+
+        Args:
+            data (torch.Tensor): Batch of sequences with shape (batch_size, length)
+            length (list): List of text lengths before padding.
+
+        Returns:
+            data (torch.Tensor): Sequences sorted by lengths in descending order.
+            length (torch.Tensor): Lengths sorted in descending order.
+            indices (torch.Tensor): The indexes of the elements in the original data.
+        """
+        length = torch.as_tensor(length, dtype=torch.int64)
+        length, sorted_indices = torch.sort(length, descending=True)
+        sorted_indices = sorted_indices.to(data.device)
+        data = data.index_select(0, sorted_indices)
+
+        indices = torch.clone(sorted_indices)
+        indices.index_put_(tuple(sorted_indices.unsqueeze(0)),
+                           torch.where(sorted_indices >= 0)[0])
```

Comment:
Does this where really work?

>>> torch.where(ind >= 0)[0]
tensor([0, 0, 0, 0, 0, 0, 0])
>>> torch.arange(ind.size(-1))
tensor([0, 1, 2, 3, 4, 5, 6])

do they have a more readable way to generate [0, 1, 2, 3, 4, 5, 6] ??

Reply:
Yes. Both (1) and (2) give`[0, 1, 2, 3, 4]`
**(1) torch.where**
```python
ind = torch.Tensor([ 1, 3, 4, 0, 2])
torch.where(ind >= 0)[0] # device='cuda:0'
``` 
**(2) torch.arange**
```python
torch.arange(ind.size(-1))) # device='cpu'
torch.arange(ind.size(-1))).to(ind) # device='cuda:0'
```

Reply:
I see. Readability of  where(ind >= 0) is not good. Using arange is better. Is there a way to avoid the device issue?